c2rust-analyze/src/rewrite/ty.rs

```diff
@@ -613,28 +613,138 @@ pub fn gen_ty_rewrites<'tcx>(
         .fn_sig_by_hir_id(hir_id)
         .unwrap_or_else(|| panic!("expected def {:?} to be a function", ldid));
 
+    let (origin_params, input_origin_args, output_origin_args) =
+        &acx.gacx.fn_origins.fn_info[&ldid.to_def_id()];
+    let hir_generics = acx.tcx().hir().get_generics(ldid);
+
+    let predicates = acx.tcx().predicates_of(ldid);
+
+    if let Some(generics) = hir_generics {
+        let mut generics_rws: Vec<Rewrite> = vec![];
```

Comment:
This approach, particularly the handling of predicates, seems fragile and also a bit complex.  It would be better to emit a rewrite that only injects the new parameters and leaves everything else unchanged (which includes any existing predicates, as in `fn f<'a, 'b: 'a>(...)`).  We should look through the HIR generics to find the right insertion point, then emit a zero-width rewrite that adds `'h0, ..., 'hN` at that point.  The logic for this is a little nontrivial - the `Span`s are scattered across the HIR `GenericParam`s and `WherePredicate`s, and it needs to handle leading/trailing commas properly when there are no regions, no tys, or no params at all (`fn foo<>(...)` is legal) - but overall, I think it will be cleaner since we won't need to regenerate any predicates.

Functions with no parameter list at all will need to have one added, but that should be straightforward since in that case there are no predicates to worry about.

Reply:
Looks about right.  More specific comments below.